dipy/align/imaffine.py

```diff
@@ -0,0 +1,777 @@
+                             moving_grid2space, prealign)
+        del prealign  # Now we must refer to self.prealign
+
+        # Multi-resolution iterations
+        original_static_grid2space = self.static_ss.get_affine(0)
+        original_moving_grid2space = self.moving_ss.get_affine(0)
+
+        for level in range(self.levels - 1, -1, -1):
+            self.current_level = level
+            max_iter = self.level_iters[level]
+            print('Optimizing level %d [max iter: %d]' % (level, max_iter))
+
+            # Resample the smooth static image to the shape of this level
+            smooth_static = self.static_ss.get_image(level)
+            current_static_shape = self.static_ss.get_domain_shape(level)
+            current_static_grid2space = self.static_ss.get_affine(level)
+
+            current_static = aff_warp(tuple(current_static_shape),
+                                      current_static_grid2space, smooth_static,
+                                      original_static_grid2space, None, False)
+
+            # The moving image is full resolution
+            current_moving_grid2space = original_moving_grid2space
+            current_moving_spacing = self.moving_ss.get_spacing(level)
+
+            current_moving = self.moving_ss.get_image(level)
+
+            # Prepare the metric for iterations at this resolution
+            self.metric.setup(transform, current_static, current_moving,
+                              current_static_grid2space,
+                              current_moving_grid2space, self.prealign)
+
+            # Optimize this level
+            if self.options is None:
+                self.options = {'gtol': 1e-4,
+                                'disp': False}
+
+            if self.method == 'L-BFGS-B':
+                self.options['maxfun'] = max_iter
+            else:
+                self.options['maxiter'] = max_iter
+
+            if SCIPY_LESS_0_12:
+                # Older versions don't expect value and gradient from
+                # the same function
+                opt = Optimizer(self.metric.distance, self.x0,
+                                method=self.method, jac=self.metric.gradient,
+                                options=self.options)
+            else:
+                opt = Optimizer(self.metric.value_and_gradient, self.x0,
+                                method=self.method, jac=True,
+                                options=self.options)
+            xopt = opt.xopt
+
+            # Update prealign matrix with optimal parameters
+            T = self.transform.param_to_matrix(xopt)
+            self.prealign = T.dot(self.prealign)
+
+            # Start next iteration at identity
+            self.x0 = self.transform.get_identity_parameters()
+
+            # Update the metric to the current solution
+            self.metric._update_dense(xopt, False)
+        return self.prealign
+
+
+def aff_warp(static, static_grid2space, moving, moving_grid2space, transform,
```

Comment:
Warp is more commonly used for nonlinear transformations it doesn't suit here.
I would say ```transform_images``` and in that way will have agreement with the streamline registration too.